pkg/kubectl/cmd/util/cmdtools.go

```diff
@@ -0,0 +1,255 @@
+	}
+	// rather than parse the shell arguments ourselves, punt to the shell
+	shell := defaultEnvShell()
+	return append(shell, cmd), true
+}
+
+func (t CmdTool) args(path string) []string {
+	if t.Shell {
+		last := t.Args[len(t.Args)-1]
+		t.Args[len(t.Args)-1] = fmt.Sprintf("%s %q", last, path)
+	} else {
+		t.Args = append(t.Args, path)
+	}
+	return t.Args
+}
+
+// Launch opens the described or returns an error. The TTY will be protected, and
+// SIGQUIT, SIGTERM, and SIGINT will all be trapped.
+func (t CmdTool) Launch(paths []string) error {
+	cmdPrefix := make([]string, len(t.Args))
+	copy(cmdPrefix, t.Args)
+	if len(t.Args) == 0 {
+		return fmt.Errorf("no %s defined, can't open %s", t.Name, paths[0])
+	}
+	for _, path := range paths {
+		abs, err := filepath.Abs(path)
+		if err != nil {
+			return err
+		}
+		t.Args = t.args(abs)
+	}
+	cmd := exec.Command(t.Args[0], t.Args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	glog.V(5).Infof("Opening file with %s %v", t.Name, t.Args)
+	if err := (term.TTY{In: os.Stdin, TryDev: true}).Safe(cmd.Run); err != nil {
+		if err, ok := err.(*exec.Error); ok {
+			if err.Err == exec.ErrNotFound {
+				return fmt.Errorf("unable to launch the %s %q", t.Name, strings.Join(cmdPrefix, " "))
+			}
+		}
+
+		//diff command will return 1 when compare different
+		if t.Name == "diff" {
+			if _, ok := err.(*exec.ExitError); ok {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("there was a problem with the %s %q", t.Name, strings.Join(cmdPrefix, " "))
+	}
+	return nil
+}
+
+type TempFile struct {
+	Prefix string
+	Suffix string
+	Buffer io.Reader
+}
+
+// LaunchTempFile reads the provided stream into multiple temporary file in the given directory
+// and file prefix, and then invokes Launch with the path of that file. It will return
+// the contents of the first file after launch, any errors that occur, and the path of the
+// temporary file so the caller can clean it up as needed.
+func (t CmdTool) LaunchTempFile(tmpFile ...TempFile) ([]byte, string, error) {
+	var paths []string
+	for _, f := range tmpFile {
+		path, err := prepareTempFile(f.Prefix, f.Suffix, f.Buffer)
+		if err != nil {
+			return nil, path, err
+		}
+		paths = append(paths, path)
+	}
+	if err := t.Launch(paths); err != nil {
+		return nil, paths[0], err
+	}
+
+	bytes, err := ioutil.ReadFile(paths[0])
+	return bytes, paths[0], err
+}
+
+func prepareTempFile(prefix, suffix string, r io.Reader) (string, error) {
+	f, err := tempFile(prefix, suffix)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	path := f.Name()
+	if _, err := io.Copy(f, r); err != nil {
+		os.Remove(path)
+		return "", err
+	}
+	// This file descriptor needs to close so the next process (Launch) can claim it.
+	f.Close()
+	return path, nil
+}
+
+func tempFile(prefix, suffix string) (f *os.File, err error) {
+	dir := os.TempDir()
+
+	for i := 0; i < 10000; i++ {
+		name := filepath.Join(dir, prefix+randSeq(5)+suffix)
```

Comment:
There are many antipatterns in this function. The biggest one is that you're rerolling a standard library function, [ioutil.TempFile](https://golang.org/pkg/io/ioutil/#TempFile). Whenever you're doing a low-level operation like creating a temp file or directory, just do a quick search on the standard library and you'll probably find what you're looking for.